### DISCUSSION TITLE
Is slow decoding with streaming decoder expected?

### DISCUSSION THREAD

**Discussion by agne-ker** [espnet/espnet#4537] [Q&A] [answered]
Hi. I'm decoding 18.36 hours of data: mean segment length 6.55s (max: 57.14s) with the streaming decoder, using https://huggingface.co/D-Keqi/espnet_asr_train_asr_streaming_transformer_raw_en_bpe500_sp_valid.acc.ave model and egs2/tedlium2/asr1/conf/decode_asr_streaming.yaml config. I'm not using any external LM and otherwise running the decode with all of the default parameters from the template script. I'm decoding with a 16 core 3.1 GHz Intel Xeon Platinum 8000 series processor. I'm running it on SGE with nj=1 (tried it with nj=ncores, but the load average was over 200!), qsub command: 

qsub -v PATH -cwd -S /bin/bash -j y -l arch=*64* -o exp/D-Keqi/espnet_asr_train_asr_streaming_transformer_raw_en_bpe500_sp_valid.acc.ave/decode_asr_streaming_asr_model_valid.acc.ave/test/logdir/q/asr_inference.log -v MKL_NUM_THREADS=1 -l h_vmem=4G   -t 1:1 exp/D-Keqi/espnet_asr_train_asr_streaming_transformer_raw_en_bpe500_sp_valid.acc.ave/decode_asr_streaming_asr_model_valid.acc.ave/test/logdir/q/asr_inference.sh. 

The SGE reported CPU time was 950522.370s (over 264 hours) and the wallclock time was 59537s (16.54 hours). This seems quite slow to me and is below the RTF reported in the "Streaming Transformer ASR with Blockwise Synchronous Beam Search" paper (https://arxiv.org/pdf/2006.14941.pdf). Could you advise whether the decoding speed I'm seeing is to be expected or if it should be faster? If so, could you please advise what should be changed to make it run faster?

**Comment by sw005320** [comment]
@D-Keqi, can you answer it for me?

**Reply by D-Keqi** [reply]
> @D-Keqi, can you answer it for me?

Sure!

**Comment by D-Keqi** [comment, marked as answer]
> Hi. I'm decoding 18.36 hours of data: mean segment length 6.55s (max: 57.14s) with the streaming decoder, using https://huggingface.co/D-Keqi/espnet_asr_train_asr_streaming_transformer_raw_en_bpe500_sp_valid.acc.ave model and egs2/tedlium2/asr1/conf/decode_asr_streaming.yaml config. I'm not using any external LM and otherwise running the decode with all of the default parameters from the template script. I'm decoding with a 16 core 3.1 GHz Intel Xeon Platinum 8000 series processor. I'm running it on SGE with nj=1 (tried it with nj=ncores, but the load average was over 200!), qsub command:
> 
> qsub -v PATH -cwd -S /bin/bash -j y -l arch=_64_ -o exp/D-Keqi/espnet_asr_train_asr_streaming_transformer_raw_en_bpe500_sp_valid.acc.ave/decode_asr_streaming_asr_model_valid.acc.ave/test/logdir/q/asr_inference.log -v MKL_NUM_THREADS=1 -l h_vmem=4G -t 1:1 exp/D-Keqi/espnet_asr_train_asr_streaming_transformer_raw_en_bpe500_sp_valid.acc.ave/decode_asr_streaming_asr_model_valid.acc.ave/test/logdir/q/asr_inference.sh.
> 
> The SGE reported CPU time was 950522.370s (over 264 hours) and the wallclock time was 59537s (16.54 hours). This seems quite slow to me and is below the RTF reported in the "Streaming Transformer ASR with Blockwise Synchronous Beam Search" paper (https://arxiv.org/pdf/2006.14941.pdf). Could you advise whether the decoding speed I'm seeing is to be expected or if it should be faster? If so, could you please advise what should be changed to make it run faster?

Hi! Thanks for your response!
If my memory is right, the experiment in "Streaming Transformer ASR with Blockwise Synchronous Beam Search" paper may be based on C++ implementation (not very sure). But I think the results you get are still a little bit slow. What is the recognition performance you get?

**Reply by agne-ker** [reply]
Thanks for getting back to me. Yes, I can see that the paper states the results reported were based on the C++ implementation. The results I get on my data (mismatched to the domain the model was trained on) is 37.1% WER and 20.9% CER.

**Reply by D-Keqi** [reply]
OK, thanks! Could you try to decode it again without using CTC prefix score (i.e. setting ctc weight to 0 during inference)? The only reason I can think of might be because if you have a lot of very long speeches (like close to 57s), it might take a long time for the model to calculate the ctc prefix score.

**Reply by agne-ker** [reply]
Thanks. I'm doing that now, but the output looks quite broken (see the below output from the logs) - is this to be expected?

▁we▁also▁have▁the▁ability▁to▁precisely▁incorporate▁various▁drug▁balidies▁into▁the▁luminity▁exercise▁where▁we▁can▁protect▁it▁as▁well▁as▁used▁the▁intrinsic▁properties▁of▁extrinsic▁properties▁of▁exertise▁of▁exertise▁of▁exertise▁of▁exertise▁of▁exertise▁of▁exerties▁of▁exertise▁of▁exerties▁of▁exertise▁and▁exertise▁of▁exertise▁of▁exertise▁and▁exertise▁of▁exertise▁of▁exertise▁where▁we▁can▁propertect▁it▁as▁well▁as▁used▁the▁intrinsic▁propertective▁as▁well▁as▁used▁as▁well▁as▁used▁the▁intrinsic▁propertities▁of▁the▁intrinsic▁propertities▁of▁those▁paycide▁of▁plasmons

**Reply by D-Keqi** [reply]
Yes, I think so when the performance is very poor.

**Reply by agne-ker** [reply]
Hmm, so is the case that the only way to improve speed without sacrificing performance is to shorten the segmentation of the speech?

**Reply by D-Keqi** [reply]
From my point of view, if the performance is ok, shortening the segmentation may be useful to improve the speed. However, if the performance is too poor due to mismatch, the Block Boundary Detection (https://arxiv.org/pdf/2006.14941.pdf) might fail, which means that the transcribed text is too long (like your example) thus consuming too much time on inference.

**Reply by agne-ker** [reply]
Thanks for your answers.
